keras_core/backend/torch/numpy.py

```diff
@@ -676,7 +677,8 @@ def pad(x, pad_width, mode="constant"):
         mode = "replicate"
     if mode != "constant" and x.ndim < 3:
         new_dims = [1] * (3 - x.ndim)
-        x = cast(x, torch.float32) if x.dtype == torch.int else x
+        if x.dtype not in (torch.float32, torch.float64):
```

Comment:
What about float16?

Reply:
I have tried float16 with `mode=reflect` and it is not supported by torch

```bash
>>> x = torch.randn(3, 10, 10, dtype=torch.float16)
>>> torch.nn.functional.pad(x, [1, 1], mode="reflect")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: "reflection_pad1d" not implemented for 'Half'
```

I believe the solution in this PR should be the same as the official one (torchvision)

https://github.com/pytorch/vision/blob/main/torchvision/transforms/_functional_tensor.py#L418-L423

Reply:
@fchollet 
Should we restore the original dtype after the op? The current code cast to float32 for non-constant padding mode.

Reply:
> Should we restore the original dtype after the op?

In fact, yes -- that is the behavior that other backends follow. Does the unit test only check float32?

Reply:
> In fact, yes -- that is the behavior that other backends follow. Does the unit test only check float32?

Actually, the unit test only checked int64.

https://github.com/keras-team/keras-core/blob/fa547ecc1a221d81def22825b6f958d7f57a5040/keras_core/ops/numpy_test.py#L2936-L2945

Please see the new comment below